### PR TITLE
Deprecate legacy input parameter construction

### PR DESCRIPTION
**Pull request by loganharbour**
Refs #19439. Prereq for #19440

### CODE REVIEW

**Comment by loganharbour** [comment]
@andrsd, I'm looking for a review on the single commit https://github.com/idaholab/moose/pull/19448/commits/da3487f9749b413d2e6096eb5e64eedb4a1b8e6e

File: modules/doc/content/newsletter/2021_11.md [hunk 1]
```diff
@@ -39,6 +39,76 @@ moose-petsc              3.15.1             build_3
...
+
+The legacy input parameter construction method has been supported for
+nearly two years. It is now slated for deprecation on Mar 1, 2022.
+After said date, any objects whose parameter construction is not
+updated will not compile.
+
+Previously, input parameters were defined with a template specialization
+of `InputParameters validParams<T>()`.
+
+This required a forward declaration in the header for an object as follows:
```
Comment (cticenhour) [inline modules/doc/content/newsletter/2021_11.md:52]:
```suggestion
Previously, input parameters were defined with a template specialization
of `InputParameters validParams<T>()`, which required a forward declaration in the object header as follows:
```

File: modules/doc/content/newsletter/2021_11.md [hunk 2]
```diff
@@ -39,6 +39,76 @@ moose-petsc              3.15.1             build_3
...
+  InputParameters params = validParams<SomeParentMooseObject>();
+  params.addParam<bool>("foo", false, "Some description");
+  return params;
+}
+```
+
+The new method of input parameter construction involves defining
+a static method on the object itself: `static InputParameters validParams()`.
+The transition of objects to the new input parameter construction method
+is straight-forward (taking the example object above). First, replace all
```
Comment (cticenhour) [inline modules/doc/content/newsletter/2021_11.md:77]:
```suggestion
is straight-forward and will be demonstrated using the example object above. First, replace all
```

**Review by cticenhour — DISMISSED** [review]
Just a couple suggestions for the newsletter entry. Content of that looks good overall.

File: modules/doc/content/newsletter/2021_11.md [hunk 3]
```diff
@@ -39,6 +39,77 @@ moose-petsc              3.15.1             build_3
...
+The legacy input parameter construction method has been supported for
+nearly two years. It is now slated for deprecation on Mar 1, 2022.
+After said date, any objects whose parameter construction is not
+updated will not compile.
+
+Previously, input parameters were defined with a template specialization
+of `InputParameters validParams<T>()`, which required a forward declaration
+in the object header as follows:
+
+This required a forward declaration in the header for an object as follows:
```
Comment (cticenhour) [inline modules/doc/content/newsletter/2021_11.md:53]:
```suggestion
```

This got left in when you accepted my review.

**Review by permcody — DISMISSED** [review]
Looks good!

**Comment by loganharbour** [comment]
I still need to change the error to a warning and fixup a few more apps. Just wanted a quick check. Thanks

File: modules/doc/content/newsletter/2021_11.md [hunk 4]
```diff
@@ -39,6 +39,75 @@ moose-petsc              3.15.1             build_3
...
+
+  static InputParameters validParams();
+
+  // continued...
+};
+```
+
+In the source of each object, replace the template specialization of
+`InputParameters validParams<T>()` with the
+`T::InputParameters validParams()` method:
```
Comment (loganharbour) [inline modules/doc/content/newsletter/2021_11.md:99]:
```suggestion
`InputParameters T::validParams()` method:
```

File: framework/include/utils/InputParameters.h [hunk 5]
```diff
@@ -1635,25 +1628,72 @@ InputParameters::get(const std::string & param1, const std::string & param2) con
...
-  // has *only* the new style validParams.
+  // the new function
+  // The SFINAE for toggling between old and new-style templating in
+  // callValidParams() will always see this function and call it even if
+  // an object has *only* the new style validParams. Therefore, we
+  // set _called_legacy_params to false so that we can avoid throwing
+  // a warning for using legacy parameters even though callValidParams()
+  // will result in calling validParams<T>. This is in support of #19439.
+  auto params = T::validParams();
+  params.template addPrivateParam<bool>("_called_legacy_params", false);
```
Comment (GiudGiud) [inline framework/include/utils/InputParameters.h:1655]: will be so nice to get rid of all that convolution when they are gone gone

File: framework/include/utils/MooseTypes.h [hunk 6]
```diff
@@ -452,20 +452,6 @@ using GenericVariableGradient = typename Moose::GenericType<VariableGradient, is
 template <bool is_ad>
 using GenericVariableSecond = typename Moose::GenericType<VariableSecond, is_ad>;
 
-#define declareADValidParams(ADObjectType)                                                         \
```
Comment (GiudGiud) [inline framework/include/utils/MooseTypes.h:455]: Pgh was using that one I think

This is effectively instant deprecation. I think we will be fine but hard to know, not all apps are in our test suite though.

**Review by GiudGiud — APPROVED** [review]
reviewed last 4 commits. This is going to make life easier when reviewing external contributions, which often include the legacy params. So this is great!!

**Pull request merged**